enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - "@dgoodwin"
+  - "@crawford"
+  - "@sdodson"
+  - "@wking"
+approvers:
+  - "@crawford"
+  - "@sdodson"
+creation-date: 2020-06-04
+last-updated: 2020-09-02
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of OpenShift 4 is to make upgrades seamless and risk free.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give the operators ability to run some checks before the upgrade to find out if the new incoming version is compatible with the existing version.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before going through the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered check if upgrade to the target version is going to be successfully without doing an upgrade.
+Use preflight checks to improve the success rate of upgrades.
+
+
+### Goals
+
+* This enhancement proposes a framework for preflight checks and discusses the CVO side implementation of that framework.
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+* Downloading and verification of images associated with release payload should be part of preflight check.
+
+### Non-Goals
+
+* This enhancement does not commit particular operators for providing preflight implementations or discuss the operator-specific logic that should be executed by preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+* It is not the responsibility of preflight checks to protect against upgrades involving more than one minor version i.e. 4.y.z to 4.(y+2).z.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The bin path need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+/bin/preflight must be executable. It can be symbolic link to the operator binary or a script or It can be non existent.
+Users can skip or override the preflight checks with `--force`.
+
+### User Stories
+
+#### Story 1
```

Comment:
I don't think this use case alone justifies a generic code mechanism, and so would either like to see a description of 2-3 real other pressing problems only solvable using this mechanism, or a description of compelling future problem classes than need it.

Right now this is pretty close to YAGNI.

Reply:
This is another good use case: https://github.com/openshift/enhancements/pull/363/files#r482442200. @LalatenduMohanty, would you mind adding it to the enhancement?

In addition to that one, the MCO could make use of these to help with the migration off of v2 Ignition configs (we are moving to v3). This is a similar situation to the one described by the Cloud Credential Operator. We can technically make use of `Upgradable=false`, but that requires adding the check one release before we are going to remove it. It's certainly possible, but it's a whole lot simpler to add the check at the same time (and in the same place) that we remove support for v2.

Reply:
@crawford Added these user cases in https://github.com/openshift/enhancements/pull/363/commits/c3340167a017e617f53a6cf1eceab01c80201a4f

Reply:
> We can technically make use of `Upgradable=false`, but that requires adding the check one release before we are going to remove it.

[Several][1] [threads][2] around motivation.  Are we collapsing to this one?  Here's another argument for preflights over `Upgradeable=False`:

1. Cluster is running release A.
2. Cluster begins updating to B, but has not yet updated operator O.
3. Admin retargets cluster to release C, but there are some AB-vs-C incompatibilities for operator O.

In a "we backport `Upgradeable=False` checks" world, it's possible that A's O hasn't learned about the check, but B's O has.  However, B's O was not yet in place to perform the check, so the unstable update to C was not blocked.  So the code performing the check may not be from a node attached to the recommended B->C update edge.  Blocking the edge from A->C would not be sufficient to ensure folks hit B's O before leaping over to C.

In a preflight world, the checking code (in C) is always a node attached to the recommended update edge.  We are isolated from any version-skew due to a heterogenous in-cluster operators.

Confirming the versions we use when requesting update recommendations, we use `optr.release.Version` [here][3], and set that [when we load our release from disk][4], so my understanding is that we will immediately begin requesting recommendations from B as soon as we load the CVO from the B release image to begin managing the update to B.

[1]: https://github.com/openshift/enhancements/pull/363#discussion_r482091423
[2]: https://github.com/openshift/enhancements/pull/363#discussion_r482093753
[3]: https://github.com/openshift/cluster-version-operator/blob/f9f5591625df130b4ec1f6c42c684877e930df5d/pkg/cvo/availableupdates.go#L51
[4]: https://github.com/openshift/cluster-version-operator/blob/f9f5591625df130b4ec1f6c42c684877e930df5d/pkg/cvo/cvo.go#L229

Reply:
@wking Thats a good point. Going to include it. 

Also I think preflight checks will be useful for rollbacks as compared to `upgradeable=false`. Though we do not support rollbacks but it would be good to know where rollbacks will part of the design for any operator.

Reply:
Fixed in https://github.com/openshift/enhancements/commit/b64620adf4e96a5d984129d2ec4edb6ce8d70ef2

Reply:
I am extremely unconvinced this proposal is superior to upgradeable=false and backport in all 4 use cases described.  The scenario here is complicated and can be addressed in simpler ways.  Customer would have had to force upgrade for this, so we should have already told them what version to go to.  And in general, we would never encourage upgrading to a new z and then suddenly jumping to a new y

Reply:
> Customer would have had to force upgrade for this, so we should have already told them what version to go to.

Wait, who has to force an update?

> And in general, we would never encourage upgrading to a new z and then suddenly jumping to a new y

What is the guard for this?  Should we only populate ClusterVersion's `status.availableUpdates` when we are in reconciliation mode?  That avoids "it's dicey to jump into a new update ->C from some undefined point in the middle of A->B", but it exposes you to "I'm hung in A->B, and need a fix delivered in C to unstick my update".  The CVO today has no such guards in this space, and will happily recommend C to you if the Cincinnati graph recommends B->C regardless of how far your cluster is along A->B.